File: cpp/p12-integer-to-roman.cpp
#import <iostream>

class Solution {
public:
    std::string intToRoman(int num)
    {
        struct IntString {
            int i;
            std::string s;
        };
        const IntString intStrings[] = {
            { 1000, "M" },
            { 900, "CM" },
            { 500, "D" },
            { 400, "CD" },
            { 100, "C" },
            { 90, "XC" },
            { 50, "L" },
            { 40, "XL" },
            { 10, "X" },
            { 9, "IX" },
            { 5, "V" },
            { 4, "IV" },
            { 3, "III" },
            { 2, "II" },
            { 1, "I" },
        };

        std::string roman;
        while (num != 0) {
            for (const auto& intString : intStrings) {
                if (num >= intString.i) {
                    roman.append(intString.s);
                    num -= intString.i;
                    break;
                }
            }
        }
        return roman;
    }
};

int main(int argc, char* argv[])
{
    struct TestCase {
        int num;
        std::string exp;
    };
    const TestCase testCases[] = {
        { 1, "I" },
        { 2, "II" },
        { 3, "III" },
        { 4, "IV" },
        { 5, "V" },
        { 6, "VI" },
        { 7, "VII" },
        { 8, "VIII" },
        { 9, "IX" },
        { 10, "X" },
        { 40, "XL" },
        { 50, "L" },
        { 58, "LVIII" },
        { 90, "XC" },
        { 91, "XCI" },
        { 94, "XCIV" },
        { 99, "XCIX" },
        { 100, "C" },
        { 400, "CD" },
        { 500, "D" },
        { 900, "CM" },
        { 1000, "M" },
        { 1994, "MCMXCIV" },
        { 3999, "MMMCMXCIX" },
    };

    Solution s;
    for (const auto& tc : testCases) {
        const auto ans = s.intToRoman(tc.num);
        if (tc.exp != ans) {
            std::cout << "FAIL. num: " << tc.num
                      << ", exp: " << tc.exp
                      << ", ans: " << ans << "\n";
        }
    }
    return 0;
}

File: py/p12-integer-to-roman.py
import collections


class Solution:
    def __init__(self):
        self.d = collections.OrderedDict()
        self.d[1000] = 'M'
        self.d[900] = 'CM'
        self.d[500] = 'D'
        self.d[400] = 'CD'
        self.d[100] = 'C'
        self.d[90] = 'XC'
        self.d[50] = 'L'
        self.d[40] = 'XL'
        self.d[10] = 'X'
        self.d[9] = 'IX'
        self.d[5] = 'V'
        self.d[4] = 'IV'
        self.d[3] = 'III'
        self.d[2] = 'II'
        self.d[1] = 'I'

    def intToRoman(self, num: int) -> str:
        roman = ''
        while num > 0:
            for key, value in self.d.items():
                if num >= key:
                    roman += value
                    num -= key
                    break
        return roman


if __name__ == '__main__':

    s = Solution()
    tests = (
        (1, 'I'),
        (2, 'II'),
        (3, 'III'),
        (4, 'IV'),
        (5, 'V'),
        (6, 'VI'),
        (7, 'VII'),
        (8, 'VIII'),
        (9, 'IX'),
        (10, 'X'),
        (40, 'XL'),
        (50, 'L'),
        (58, 'LVIII'),
        (90, 'XC'),
        (91, 'XCI'),
        (94, 'XCIV'),
        (99, 'XCIX'),
        (100, 'C'),
        (400, 'CD'),
        (500, 'D'),
        (900, 'CM'),
        (1000, 'M'),
        (1994, 'MCMXCIV'),
        (3999, 'MMMCMXCIX'),
    )

    for test in tests:

        num = test[0]
        exp = test[1]
        ans = s.intToRoman(num)
        if exp != ans:
            print(f'FAIL. num: {num}, exp: {exp}, ans: {ans}')
